Sources/NIOExtras/WritePCAPHandler.swift

```diff
@@ -0,0 +1,649 @@
+/// data as synthesized TCP packet captures.
+///
+/// You will be able to open the `.pcap` file in for example [Wireshark](https://www.wireshark.org) or
+/// [`tcpdump`](http://www.tcpdump.org). Using `NIOWritePCAPHandler` to write your `.pcap` files can be useful for
+/// example when your real network traffic is TLS protected (so `tcpdump`/Wireshark can't read it directly), or if you
+/// don't have enough privileges on the running host to dump the network traffic.
+///
+/// `NIOWritePCAPHandler` will also work with Unix Domain Sockets in which case it will still synthesize a TCP packet
+/// capture with local address `111.111.111.111` (port `1111`) and remote address `222.222.222.222` (port `2222`).
+public class NIOWritePCAPHandler {
+    public enum Mode {
+        case client
+        case server
+    }
+
+    private enum CloseState {
+        case notClosing
+        case closedInitiatorLocal
+        case closedInitiatorRemote
+    }
+    
+    private let fileSink: (ByteBuffer) -> Void
+    private let mode: Mode
+    private let maxPayloadSize = Int(UInt16.max - 40 /* needs to fit into the IPv4 header which adds 40 */)
+    private var buffer: ByteBuffer!
+    private var readInboundBytes = 0
+    private var writtenOutboundBytes = 0
+    private var closeState = CloseState.notClosing
+    
+    private static let fakeLocalAddress = try! SocketAddress(ipAddress: "111.111.111.111", port: 1111)
+    private static let fakeRemoteAddress = try! SocketAddress(ipAddress: "222.222.222.222", port: 2222)
+    
+    private var localAddress: SocketAddress?
+    private var remoteAddress: SocketAddress?
+
+    public static var pcapFileHeader: ByteBuffer {
+        var buffer = ByteBufferAllocator().buffer(capacity: 24)
+        buffer.writePCAPHeader()
+        return buffer
+    }
+
+    /// Initialize a `NIOWritePCAPHandler`.
+    ///
+    /// - parameters:
+    ///     - fakeLocalAddress: Allows you to optionally override the local address to be different from the real one.
+    ///     - fakeRemoteAddress: Allows you to optionally override the remote address to be different from the real one.
+    ///     - fileSink: The `fileSink` closure is called every time a new chunk of the `.pcap` file is ready to be
+    ///                 written to disk or elsewhere. See `NIOSynchronizedFileSink` for a convenient way to write to
+    ///                 disk.
+    public init(mode: Mode,
+                fakeLocalAddress: SocketAddress? = nil,
+                fakeRemoteAddress: SocketAddress? = nil,
+                fileSink: @escaping (ByteBuffer) -> Void) {
+        self.fileSink = fileSink
+        self.mode = mode
+        if let fakeLocalAddress = fakeLocalAddress {
+            self.localAddress = fakeLocalAddress
+        }
+        if let fakeRemoteAddress = fakeRemoteAddress {
+            self.remoteAddress = fakeRemoteAddress
+        }
+    }
+    
+    private func writeBuffer(_ buffer: ByteBuffer, context: ChannelHandlerContext) {
```

Comment:
This doesn't seem to need the `ChannelHandlerContext` anymore.

Reply:
done